internal/pkg/encryption/keys/kms.go

```diff
@@ -0,0 +1,165 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package keys
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/luks"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/token"
+	"github.com/siderolabs/go-retry/retry"
+	"github.com/siderolabs/kms-client/api/kms"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/siderolabs/talos/internal/pkg/endpoint"
+	"github.com/siderolabs/talos/internal/pkg/smbios"
+)
+
+// KMSToken is the userdata stored in the partition token metadata.
+type KMSToken struct {
+	SealedData []byte `json:"sealedData"`
+}
+
+// KMSKeyHandler seals token using KMS service.
+type KMSKeyHandler struct {
+	key
+	kmsEndpoint string
+}
+
+// NewKMSKeyHandler creates new KMSKeyHandler.
+func NewKMSKeyHandler(key key, kmsEndpoint string) (*KMSKeyHandler, error) {
+	return &KMSKeyHandler{
+		key:         key,
+		kmsEndpoint: kmsEndpoint,
+	}, nil
+}
+
+// NewKey implements Handler interface.
+func (h *KMSKeyHandler) NewKey() (*encryption.Key, token.Token, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	conn, err := h.getConn(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error dialing KMS endpoint %q: %w", h.kmsEndpoint, err)
+	}
+
+	client := kms.NewKMSServiceClient(conn)
+
+	s, err := smbios.GetSMBIOSInfo()
```

Comment:
this is already populated in the systeminformation controller

Reply:
we don't have access to the controllers here.

Reply:
I think this is a really valid concern - if we want to make anything in Talos to fix "zero" UUIDs, we need to make sure we fetch UUID from a single place everywhere. we don't need to fetch UUID here, we could fetch it before hitting the encryption code and pass it in as a parameter

Reply:
Ok, I've made a change.